deeppavlov/core/trainers/fit_trainer.py

```diff
@@ -0,0 +1,260 @@
+                c_out = component_config['out']
+                in_y = component_config.get('in_y', None)
+                main = component_config.get('main', False)
+                self._chainer.append(component, c_in, c_out, in_y, main)
+        self._built = True
+
+    def _load(self) -> None:
+        if not self._loaded:
+            self._chainer.destroy()
+            self._chainer = build_model({'chainer': self.chainer_config}, load_trained=self._saved)
+            self._loaded = True
+
+    def get_chainer(self) -> Chainer:
+        """Return a :class:`~deeppavlov.core.common.chainer.Chainer` built from ``self.chainer_config`` for inference"""
+        self._load()
+        return self._chainer
+
+    def train(self, iterator: Union[DataFittingIterator, DataLearningIterator]) -> None:
+        """Call :meth:`~fit_chainer` with provided data iterator as an argument"""
+        self.fit_chainer(iterator)
+        self._saved = True
+
+    def test(self, data: Iterable[Tuple[Collection[Any], Collection[Any]]],
+             metrics: Optional[Collection[Metric]] = None, *,
+             start_time: Optional[float] = None, show_examples: Optional[bool] = None) -> dict:
+        """
+        Calculate metrics and return reports on provided data for currently stored
+        :class:`~deeppavlov.core.common.chainer.Chainer`
+
+        Args:
+            data: iterable of batches of inputs and expected outputs
+            metrics: collection of metrics namedtuples containing names for report, metric functions
+                and their inputs names (if omitted, ``self.metrics`` is used)
+            start_time: start time for test report
+            show_examples: a flag used to return inputs, expected outputs and predicted outputs for the last batch
+                in a result report (if omitted, ``self.show_examples`` is used)
+
+        Returns:
+            a report dict containing calculated metrics, spent time value, examples count in tested data
+            and maybe examples
+        """
+
+        if start_time is None:
+            start_time = time.time()
+        if show_examples is None:
+            show_examples = self.show_examples
+        if metrics is None:
+            metrics = self.metrics
+
+        expected_outputs = list(set().union(self._chainer.out_params, *[m.inputs for m in metrics]))
+
+        outputs = {out: [] for out in expected_outputs}
+        examples = 0
+
+        data = islice(data, self.max_test_batches)
+
+        for x, y_true in data:
+            examples += len(x)
+            y_predicted = list(self._chainer.compute(list(x), list(y_true), targets=expected_outputs))
+            if len(expected_outputs) == 1:
+                y_predicted = [y_predicted]
+            for out, val in zip(outputs.values(), y_predicted):
+                out += list(val)
+
+        if examples == 0:
+            return {'eval_examples_count': 0, 'metrics': None, 'time_spent': str(datetime.timedelta(seconds=0))}
```

Comment:
Should we raise warning or even error here?

Reply:
log.warning('Got empty data iterable for scoring')